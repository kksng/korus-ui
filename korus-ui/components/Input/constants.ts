export const predefinedAllowedSymbols = {
  numbers: /\d/,
};

export const predefinedForbiddenSymbols = {
  numbers: /\d/,
};
