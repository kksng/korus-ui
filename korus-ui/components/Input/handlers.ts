import * as React from 'react';

import { isFunction, isNil } from 'lodash';
import { CustomEventHandler, SetState, AdjustCursor } from '../../commonTypes';
import { InputProps } from './types';
import {
  isSymbolAllowed, isSymbolForbidden, transformToCase, getSelection, getNewValueLength, getMaxPastedLength, getNewPastedValue,
} from './helpers';
import { stringToMaxLength } from '../../utils';

/**
 * Function creates change event handler
 * @param {InputProps} props - properties of component
 * @param {SetState<string>} setValue - set state action
 * @param {AdjustCursor} adjustCursor - function sets cursor to specified or previous position
 *
 * @returns {CustomEventHandler<React.ChangeEvent<HTMLInputElement>>} - change event handler
 */
export const createChangeHandler = (
  props: InputProps,
  setValue: SetState<string>,
  adjustCursor: AdjustCursor,
): CustomEventHandler<React.ChangeEvent<HTMLInputElement>> => (event) => {
  const { value } = event.target;
  const {
    allowedSymbols, forbiddenSymbols, letterCase, maxLength,
  } = props;

  if (isSymbolForbidden(value, forbiddenSymbols)) return;

  if (!isSymbolAllowed(value, allowedSymbols)) return;

  // Next block prevents from replacing previous value if maxLength is exceeded
  // and sets cursor back to initial position
  if (!isNil(maxLength) && value.length > maxLength) {
    adjustCursor(event);
    return;
  }

  const maxLengthAdjustedValue = stringToMaxLength(value, maxLength);

  const newValue = letterCase ? transformToCase(value, letterCase) : maxLengthAdjustedValue;

  if (props.value === undefined) {
    setValue(newValue);
  }

  props.onChange?.({
    ...event,
    component: {
      name: props.name,
      value: newValue,
    },
  });
};

/**
 * Function creates clear event handler
 * @param {InputProps} props - properties of component
 * @param {SetState<string>} setValue - set state action
 * @param {React.MutableRefObject<HTMLInputElement | null>} inputRef - input ref
 *
 * @returns {CustomEventHandler<React.MouseEvent<HTMLInputElement>>} - clear event handler
 */
export const createClearHandler = (
  props: InputProps,
  setValue: SetState<string>,
  inputRef: React.MutableRefObject<HTMLInputElement | null>,
): CustomEventHandler<React.MouseEvent<HTMLInputElement>> => (event) => {
  event.preventDefault();

  if (inputRef.current) inputRef.current.focus();

  if (props.value === undefined) {
    setValue('');
  }

  props.onChange?.({
    ...event,
    component: {
      name: props.name,
      value: '',
    },
  });
};

/**
 * Function creates blur event handler
 * @param {InputProps} props - properties of component
 * @param {SetState<boolean>} setFocused - set state action
 * @param {() => boolean} validate - validation function
 *
 * @returns {React.FocusEventHandler<HTMLInputElement>} - blur event handler
 */
export const createBlurHandler = (
  props: InputProps,
  setFocused: SetState<boolean>,
  validate: () => boolean,
): React.FocusEventHandler<HTMLInputElement> => (event) => {
  setFocused(false);

  const newValid = validate();

  props.onBlur?.({
    ...event,
    component: {
      isValid: newValid,
      name: props.name,
      value: event.target.value,
    },
  });
};

/**
 * Function creates focus event handler
 * @param {InputProps} props - properties of component
 * @param {boolean} isValid - flag defines if input value is valid
 * @param {SetState<boolean>} setFocused - set state action
 *
 * @returns {React.FocusEventHandler<HTMLInputElement>} - focus event handler
 */
export const createFocusHandler = (
  props: InputProps,
  isValid: boolean,
  setFocused: SetState<boolean>,
): React.FocusEventHandler<HTMLInputElement> => (event) => {
  setFocused(true);

  props.onFocus?.({
    ...event,
    component: {
      isValid,
      name: props.name,
      value: event.target.value,
    },
  });
};

/**
 * Function creates key down event handler
 * @param {InputProps} props - properties of component
 *
 * @returns {React.KeyboardEventHandler<HTMLInputElement>} - key down event handler
 */
export const createKeyDownHandler = (
  props: InputProps,
): React.KeyboardEventHandler<HTMLInputElement> => (event) => {
  if (event.key === 'Enter') {
    props.onEnterPress?.({
      ...event,
      component: {
        name: props.name,
        value: event.currentTarget.value,
      },
    });
  }
};

/**
 * Function creates reset handler
 * @param {InputProps} props - properties of component
 * @param {SetState<string>} setValue - set state action
 *
 * @returns {() => void} - reset handler
 */
export const createResetHandler = (
  props: InputProps,
  setValue: SetState<string>,
) => () => {
  const newValue = props.defaultValue || '';

  setValue(newValue);

  props.onChange?.({
    component: {
      name: props.name,
      value: newValue,
    },
  });
};

/**
 * Function creates paste event handler which takes control if new value exceeds maxLength
 *
 * @param {InputProps} props - properties of component
 * @param {string} value - current value of input
 * @param {SetState<string>} setValue - set state function for input value
 * @param {AdjustCursor} adjustCursor - function sets cursor to specified or previous position if maxLength is exceeded
 *
 * @returns {CustomEventHandler<React.ClipboardEvent<HTMLInputElement>>} - paste event handler
 */
export const createPasteHandler = (
  props: InputProps,
  value: string,
  setValue: SetState<string>,
  adjustCursor: AdjustCursor,
): CustomEventHandler<React.ClipboardEvent<HTMLInputElement>> => (event) => {
  const {
    maxLength, letterCase, isDisabled, onChange, name,
  } = props;
  const inputElement = event.target as HTMLInputElement;

  if (isDisabled) return;

  const { selectionStart, selectionEnd, selectedRange } = getSelection(inputElement);
  const pastedValue = event.clipboardData.getData('Text');
  const newValueLength = getNewValueLength(value, pastedValue, selectedRange);

  if (!isNil(maxLength) && newValueLength > maxLength) {
    const maxPastedLength = getMaxPastedLength(value, maxLength, selectedRange);

    const adjustedPastedValue = stringToMaxLength(pastedValue, maxPastedLength);

    const newValue = getNewPastedValue({
      adjustedPastedValue,
      maxLength,
      oldValue: value,
      selectedRange,
      selectionEnd,
      selectionStart,
    });

    if (newValue !== null) {
      event.preventDefault();

      const adjustedNewValue = letterCase ? transformToCase(newValue, letterCase) : newValue;

      if (props.value === undefined) {
        setValue(adjustedNewValue);
      }

      if (isFunction(onChange)) {
        const customEvent = {
          ...event,
          component: {
            name,
            value: adjustedNewValue,
          },
          target: event.target as HTMLInputElement,
        };

        onChange(customEvent);
      }

      if (selectionStart !== null) adjustCursor(event, selectionStart + maxPastedLength);
    } else {
      adjustCursor(event);
    }
  }
};
