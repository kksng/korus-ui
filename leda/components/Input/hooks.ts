import * as React from 'react';

import {
  useRunAfterUpdate,
} from '~/utils';

/**
 * Hook provides function that returns cursor to initial position after cancelled paste or change event
 *
 * @returns {(event: React.ChangeEvent<HTMLInputElement> | React.ClipboardEvent<HTMLInputElement>) => void}
 */
export const useSetCursorToPrevPosition = (): (event: React.ChangeEvent<HTMLInputElement> | React.ClipboardEvent<HTMLInputElement>) => void => {
  const [isPasteEvent, setIsPasteEvent] = React.useState(false);
  const [pasteSelectionStart, setPasteSelectionStart] = React.useState<number | null>(null);

  const runAfterUpdate = useRunAfterUpdate();

  return (event: React.ChangeEvent<HTMLInputElement> | React.ClipboardEvent<HTMLInputElement>) => {
    const inputElement = event.target as HTMLInputElement;

    if (event.type === 'paste') {
      setIsPasteEvent(true);
      setPasteSelectionStart(inputElement.selectionStart);
    }

    if (event.type === 'change') {
      setIsPasteEvent(false);
    }

    const prevCursorPositionForPaste = pasteSelectionStart || 0;
    const prevCursorPositionForChange = inputElement.selectionStart ? inputElement.selectionStart - 1 : 0;

    const prevCursorPosition = isPasteEvent ? prevCursorPositionForPaste : prevCursorPositionForChange;

    runAfterUpdate(() => {
      inputElement.selectionStart = prevCursorPosition;
      inputElement.selectionEnd = prevCursorPosition;
    });
  };
};
