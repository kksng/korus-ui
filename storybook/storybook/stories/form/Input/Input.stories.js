// @flow
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import
{
  AllowedSymbols,
  Api,
  BasicUsage,
  Customization,
  LetterCase,
  MaxLength,
} from './Examples';
import { Story } from '../../../components/Story';
import { propsDesc, validationPropsDesc } from './propsDescription';
import { ApiStory } from '../../../components/ApiStory';

const compoundProps = [
  { componentName: 'Input', props: propsDesc },
  { componentName: 'Validation', props: validationPropsDesc },
];

storiesOf('Form|Input', module)
  .add('API', context => (<ApiStory {...Api} compoundCustomProps={compoundProps} context={context} />))
  .add('Basic Usage', context => (<Story {...BasicUsage} compoundCustomProps={compoundProps} context={context} />))
  .add('Customization', context => (<Story {...Customization} compoundCustomProps={compoundProps} context={context} />));

storiesOf('Form|Input.Props', module)
  .add('allowedSymbols', context => (<Story {...AllowedSymbols} compoundCustomProps={compoundProps} context={context} />))
  .add('letterCase', context => (<Story {...LetterCase} compoundCustomProps={compoundProps} context={context} />))
  .add('maxLength', context => (<Story {...MaxLength} compoundCustomProps={compoundProps} context={context} />));
