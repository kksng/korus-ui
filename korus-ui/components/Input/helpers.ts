import { isNumber, isString } from 'lodash';
import { predefinedAllowedSymbols, predefinedForbiddenSymbols } from './constants';
import { GetNewPastedValue, GetSelection, InputProps } from './types';

export const isSymbolAllowed = (value: string, allowedSymbols?: keyof typeof predefinedAllowedSymbols | RegExp): boolean => {
  if (!allowedSymbols || value.length === 0) return true;

  if (isString(allowedSymbols)) {
    const predefinedRegExp = predefinedAllowedSymbols[allowedSymbols];

    if (!predefinedRegExp) throw new Error(`L.Input: no such predefined allowedSymbols - "${allowedSymbols}"!`);

    return [...value].every((symbol) => predefinedRegExp.test(symbol));
  }

  if (allowedSymbols instanceof RegExp) {
    const regExp: RegExp = allowedSymbols;

    return [...value].every((symbol) => regExp.test(symbol));
  }

  throw new Error('L.Input: allowedSymbols prop accepts only predefined string or RegExp!');
};

export const isSymbolForbidden = (value: string, forbiddenSymbols?: keyof typeof predefinedForbiddenSymbols | RegExp): boolean => {
  if (!forbiddenSymbols || value.length === 0) return false;

  if (isString(forbiddenSymbols)) {
    const predefinedRegExp = predefinedForbiddenSymbols[forbiddenSymbols];

    if (!predefinedRegExp) throw new Error(`L.Input: no such predefined forbiddenSymbols - "${forbiddenSymbols}"!`);

    return [...value].some((symbol) => predefinedRegExp.test(symbol));
  }

  if (forbiddenSymbols instanceof RegExp) {
    const regExp: RegExp = forbiddenSymbols;

    return [...value].some((symbol) => regExp.test(symbol));
  }

  throw new Error('L.Input: forbiddenSymbols prop accepts only predefined string or RegExp!');
};

export const transformToCase = (letter: string, letterCase: InputProps['letterCase']): string | never => {
  if (letterCase === 'lower') return letter.toLowerCase();

  if (letterCase === 'upper') return letter.toUpperCase();

  throw new Error('L.Input: letterCase prop must be \'lower\' or \'upper\'!');
};

export const getValue = (valueProp: string | null | undefined, valueState: string): string => {
  if (valueProp === undefined) return valueState;

  if (valueProp === null) {
    return '';
  }

  return valueProp;
};

/**
 * Helper defines if value was selected completely or partly
 * @param {number | null} selectedRange - selected range
 * @param {number} oldValueLength - length of previous value
 *
 * @returns {[boolean, boolean]} - array of flags that define if value was selected completely or partly
 */
export const getSelectionType = (selectedRange: number | null, oldValueLength: number): [boolean, boolean] => {
  const isAllSelected = selectedRange === oldValueLength;
  const isPartSelected = !isAllSelected && selectedRange !== 0;

  return [isAllSelected, isPartSelected];
};

/**
 * Helper defines cursor position or selection range
 */
export const getSelection: GetSelection = (inputElement) => {
  const selectedRange = isNumber(inputElement.selectionEnd) && isNumber(inputElement.selectionStart)
    ? (inputElement.selectionEnd - inputElement.selectionStart)
    : null;

  return {
    selectedRange,
    selectionEnd: inputElement.selectionEnd,
    selectionStart: inputElement.selectionStart,
  };
};

/**
 * Helper calculates new value length
 * @param {string} oldValue - value of input before paste
 * @param {pastedValue} pastedValue - pasted value
 * @param {number | null} selectedRange - selected range for paste
 *
 * @returns {number} - new value length
 */
export const getNewValueLength = (oldValue: string, pastedValue: string, selectedRange: number | null): number => {
  const [isAllSelected, isPartSelected] = getSelectionType(selectedRange, oldValue.length);

  if (isAllSelected) return pastedValue.length;
  if (isPartSelected && selectedRange) return oldValue.length - selectedRange + pastedValue.length;
  return oldValue.length + pastedValue.length;
};

/**
 * Helper calculates maximum allowed length of pasted value
 * @param {string} oldValue - value of input before paste
 * @param {number} maxLength - maximum allowed value length
 * @param {number} selectedRange - selected range for paste
 *
 * @returns {number} - maximum allowed length of pasted value
 */
export const getMaxPastedLength = (oldValue: string, maxLength: number, selectedRange: number | null): number => {
  const [isAllSelected, isPartSelected] = getSelectionType(selectedRange, oldValue.length);

  if (isAllSelected || oldValue === '') return maxLength;
  if (isPartSelected && selectedRange) return maxLength - (oldValue.length - selectedRange);
  return maxLength - oldValue.length;
};

/**
 * Helper defines new value after paste
 */
export const getNewPastedValue: GetNewPastedValue = (
  {
    adjustedPastedValue,
    selectionEnd,
    selectedRange,
    maxLength,
    selectionStart,
    oldValue,
  },
) => {
  const [isAllSelected, isPartSelected] = getSelectionType(selectedRange, oldValue.length);

  if (isAllSelected || oldValue === '') return adjustedPastedValue;
  if (isPartSelected && isNumber(selectionStart) && isNumber(selectionEnd)) return oldValue.substring(0, selectionStart) + adjustedPastedValue + oldValue.substring(selectionEnd, oldValue.length);
  if (oldValue.length < maxLength && isNumber(selectionStart)) return oldValue.substring(0, selectionStart) + adjustedPastedValue + oldValue.substring(selectionStart, oldValue.length);
  return null;
};
