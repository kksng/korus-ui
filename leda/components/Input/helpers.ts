import { isNumber, isString } from 'lodash';
import { predefinedAllowedSymbols, predefinedForbiddenSymbols } from './constants';
import { GetNewPastedValue, GetSelection, InputProps } from './types';

export const isSymbolAllowed = (value: string, allowedSymbols?: keyof typeof predefinedAllowedSymbols | RegExp): boolean => {
  if (!allowedSymbols || value.length === 0) return true;

  if (isString(allowedSymbols)) {
    const predefinedRegExp = predefinedAllowedSymbols[allowedSymbols];

    if (!predefinedRegExp) throw new Error(`L.Input: no such predefined allowedSymbols - "${allowedSymbols}"!`);

    return [...value].every((symbol) => predefinedRegExp.test(symbol));
  }

  if (allowedSymbols instanceof RegExp) {
    const regExp: RegExp = allowedSymbols;

    return [...value].every((symbol) => regExp.test(symbol));
  }

  throw new Error('L.Input: allowedSymbols prop accepts only predefined string or RegExp!');
};

export const isSymbolForbidden = (value: string, forbiddenSymbols?: keyof typeof predefinedForbiddenSymbols | RegExp): boolean => {
  if (!forbiddenSymbols || value.length === 0) return false;

  if (isString(forbiddenSymbols)) {
    const predefinedRegExp = predefinedForbiddenSymbols[forbiddenSymbols];

    if (!predefinedRegExp) throw new Error(`L.Input: no such predefined forbiddenSymbols - "${forbiddenSymbols}"!`);

    return [...value].some((symbol) => predefinedRegExp.test(symbol));
  }

  if (forbiddenSymbols instanceof RegExp) {
    const regExp: RegExp = forbiddenSymbols;

    return [...value].some((symbol) => regExp.test(symbol));
  }

  throw new Error('L.Input: forbiddenSymbols prop accepts only predefined string or RegExp!');
};

export const transformToCase = (letter: string, letterCase: InputProps['letterCase']): string | never => {
  if (letterCase === 'lower') return letter.toLowerCase();

  if (letterCase === 'upper') return letter.toUpperCase();

  throw new Error('L.Input: letterCase prop must be \'lower\' or \'upper\'!');
};

export const getValue = (valueProp: string | null | undefined, valueState: string): string => {
  if (valueProp === undefined) return valueState;

  if (valueProp === null) {
    return '';
  }

  return valueProp;
};

/**
 * Helper defines cursor position or selection range
 * @param {HTMLInputElement} inputElement - input element
 */
export const getSelection: GetSelection = (inputElement) => {
  const range = isNumber(inputElement.selectionEnd) && isNumber(inputElement.selectionStart)
    ? (inputElement.selectionEnd - inputElement.selectionStart)
    : null;

  return {
    end: inputElement.selectionEnd,
    range,
    start: inputElement.selectionStart,
  };
};

export const getNewValueLength = (oldValue: string, pastedValue: string, selectedRange: number | null): number => {
  const isAllSelected = selectedRange === oldValue.length;
  const isPartSelected = !isAllSelected && selectedRange !== 0;

  if (isAllSelected) return pastedValue.length;
  if (isPartSelected && selectedRange) return oldValue.length - selectedRange + pastedValue.length;
  return oldValue.length + pastedValue.length;
};

export const getMaxPastedLength = (oldValue: string, maxLength: number, selectedRange: number | null): number => {
  const isAllSelected = selectedRange === oldValue.length;
  const isPartSelected = !isAllSelected && selectedRange !== 0;

  if (isAllSelected || oldValue === '') return maxLength;
  if (isPartSelected && selectedRange) return maxLength - (oldValue.length - selectedRange);
  return maxLength - oldValue.length;
};

export const getNewPastedValue: GetNewPastedValue = (
  {
    adjustedPastedValue,
    selectionEnd,
    selectedRange,
    maxLength,
    selectionStart,
    oldValue,
  },
) => {
  const isAllSelected = selectedRange === oldValue.length;
  const isPartSelected = !isAllSelected && selectedRange !== 0;

  if (isAllSelected || oldValue === '') return adjustedPastedValue;
  if (isPartSelected && isNumber(selectionStart) && isNumber(selectionEnd)) return oldValue.substring(0, selectionStart) + adjustedPastedValue + oldValue.substring(selectionEnd, oldValue.length);
  if (oldValue.length < maxLength && isNumber(selectionStart)) return oldValue.substring(0, selectionStart) + adjustedPastedValue + oldValue.substring(selectionStart, oldValue.length);
  return null;
};
