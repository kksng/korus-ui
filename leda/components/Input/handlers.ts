import * as React from 'react';
import { isFunction } from 'lodash';
import { CustomEventHandler, SetState } from '../../commonTypes';
import { InputProps } from './types';
import { isSymbolAllowed, isSymbolForbidden, transformToCase } from './helpers';
import { stringToMaxLength } from '../../utils';

export const createChangeHandler = (
  props: InputProps,
  setValue: SetState<string>,
  runAfterUpdate: (callback: () => void) => void,
): CustomEventHandler<React.ChangeEvent<HTMLInputElement>> => (event) => {
  const { value } = event.target;
  const {
    allowedSymbols, forbiddenSymbols, letterCase, maxLength,
  } = props;

  if (isSymbolForbidden(value, forbiddenSymbols)) return;

  if (!isSymbolAllowed(value, allowedSymbols)) return;

  const inputElement = event.target;
  const prevSelection = {
    start: inputElement.selectionStart,
    end: inputElement.selectionEnd,
  };

  if (maxLength != null && value.length > maxLength) {
    const newSelectionPosition = prevSelection.start ? prevSelection.start - 1 : 0;
    // возвращаем курсор на предыдущую позицию в блокирующем режиме
    runAfterUpdate(() => {
      inputElement.selectionStart = newSelectionPosition;
      inputElement.selectionEnd = newSelectionPosition;
    });

    return;
  }

  const maxLengthAdjustedValue = stringToMaxLength(value, maxLength);

  const newValue = letterCase ? transformToCase(maxLengthAdjustedValue, letterCase) : maxLengthAdjustedValue;

  if (props.value === undefined) {
    setValue(newValue);
  }

  props.onChange?.({
    ...event,
    component: {
      value: newValue,
      name: props.name,
    },
  });
};

export const createClearHandler = (
  props: InputProps,
  setValue: SetState<string>,
  inputRef: React.MutableRefObject<HTMLInputElement | null>,
): CustomEventHandler<React.MouseEvent<HTMLInputElement>> => (event) => {
  event.preventDefault();

  if (inputRef.current) inputRef.current.focus();

  if (props.value === undefined) {
    setValue('');
  }

  props.onChange?.({
    ...event,
    component: {
      value: '',
      name: props.name,
    },
  });
};

export const createBlurHandler = (
  props: InputProps,
  setFocused: SetState<boolean>,
  validate: () => boolean,
): React.FocusEventHandler<HTMLInputElement> => (event) => {
  setFocused(false);

  const newValid = validate();

  props.onBlur?.({
    ...event,
    component: {
      value: event.target.value,
      name: props.name,
      isValid: newValid,
    },
  });
};

export const createFocusHandler = (
  props: InputProps,
  isValid: boolean,
  setFocused: SetState<boolean>,
): React.FocusEventHandler<HTMLInputElement> => (event) => {
  setFocused(true);

  props.onFocus?.({
    ...event,
    component: {
      value: event.target.value,
      name: props.name,
      isValid,
    },
  });
};

export const createKeyDownHandler = (
  props: InputProps,
): React.KeyboardEventHandler<HTMLInputElement> => (event) => {
  if (event.key === 'Enter') {
    props.onEnterPress?.({
      ...event,
      component: {
        value: event.currentTarget.value,
        name: props.name,
      },
    });
  }
};

export const createResetHandler = (
  props: InputProps,
  setValue: SetState<string>,
) => () => {
  const newValue = props.defaultValue || '';

  setValue(newValue);

  props.onChange?.({
    component: {
      name: props.name,
      value: newValue,
    },
  });
};

export const createPasteHandler = (
  props: InputProps,
  value: string,
  setValue: SetState<string>,
) => (event: React.ClipboardEvent<HTMLInputElement>) => {
  const {
    onChange, name, isDisabled, maxLength,
  } = props;

  event.preventDefault();
  // по неизвестным причинам onPaste работает даже на отключенных инпутах
  if (isDisabled) return;

  const text = value + event.clipboardData.getData('Text');
  const maxLengthAdjustedValue = stringToMaxLength(text, maxLength);
  const newValue = maxLength != null ? maxLengthAdjustedValue : text;

  if (props.value === undefined) {
    setValue(newValue);
  }

  if (isFunction(onChange)) {
    const customEvent = {
      ...event,
      target: event.target as HTMLInputElement,
      component: {
        name,
        value: newValue,
      },
    };

    onChange(customEvent);
  }
};
