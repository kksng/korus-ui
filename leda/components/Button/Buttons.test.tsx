// @ts-nocheck
import React from 'react';
import {
  render,
  fireEvent,
} from '@testing-library/react';
import { Button } from './index';
import { Input } from '../Input';


describe('Button snapshots collection', () => {
  test('is Button render right with basic props?', () => {
    const wrapper = render((
      <Button>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button render right with isLoading?', () => {
    const wrapper = render((
      <Button isLoading>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button render right with isDisabled attributes?', () => {
    const wrapper = render((
      <Button isDisabled>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button render right with Form and onClick props?', () => {
    const wrapper = render((
      <div>
        <Input form="test" name="in" isRequired />
        <Button onClick={jest.fn()} form="test" isDisabled>test</Button>
      </div>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
});
describe('Button condition set collection', () => {
  test('is Button attributes set right?', () => {
    const { container, rerender } = render(<Button _primary>test</Button>);

    expect(container.querySelectorAll('button.primary')[0])
      .toBeInTheDocument();

    rerender(<Button _secondary>test</Button>);

    expect(container.querySelectorAll('button.secondary')[0])
      .toBeInTheDocument();

    rerender(<Button _success>test</Button>);

    expect(container.querySelectorAll('button.success')[0])
      .toBeInTheDocument();

    rerender(<Button _warning>test</Button>);

    expect(container.querySelectorAll('button.warning')[0])
      .toBeInTheDocument();

    rerender(<Button _danger>test</Button>);

    expect(container.querySelectorAll('button.danger')[0])
      .toBeInTheDocument();

    rerender(<Button _small>test</Button>);

    expect(container.querySelectorAll('button.small')[0])
      .toBeInTheDocument();

    rerender(<Button _large>test</Button>);

    expect(container.querySelectorAll('button.large')[0])
      .toBeInTheDocument();

    rerender(<Button _block>test</Button>);

    expect(container.querySelectorAll('button.block')[0])
      .toBeInTheDocument();
  });
  test('is Button work right with isDisabled property?', () => {
    const onClickHandler = jest.fn();
    const { container } = render(<Button isDisabled onClick={onClickHandler}>test</Button>);
    const button = container.querySelectorAll('button')[0];

    expect(container.querySelectorAll('button.disabled')[0])
      .toBeInTheDocument();

    fireEvent.click(button);

    expect(onClickHandler)
      .not
      .toHaveBeenCalled();
  });
  test('is Button work right with isLoading property?', () => {
    const onClickHandler = jest.fn();
    const { container } = render(<Button isLoading onClick={onClickHandler}>test</Button>);
    const button = container.querySelectorAll('button')[0];

    expect(container.querySelectorAll('button.loading')[0])
      .toBeInTheDocument();

    fireEvent.click(button);

    expect(onClickHandler)
      .not
      .toHaveBeenCalled();
  });
});
describe('Button form validation test collection', () => {
  test('is Button work right with form attributes submit Form with valid value?', () => {
    const onSubmitValid = jest.fn();
    const onSubmitInvalid = jest.fn();
    const email = 'some@email.ru';
    const { getByText } = render((
      <div>
        <Input form="test" name="email" invalidMessage="error" isRequired validator="email" value={email} />
        <Button onClick={onSubmitValid} onValidationFail={onSubmitInvalid} form="test">test</Button>
      </div>
    ));

    fireEvent.click(getByText('test'));
    expect(onSubmitValid).toBeCalledTimes(1);
    expect(onSubmitInvalid).toBeCalledTimes(0);
  });
  test('is Button work right with form attributes submit Form with invalid value?', () => {
    const onSubmitValid = jest.fn();
    const onSubmitInvalid = jest.fn();
    const email = '-----';
    const { getByText } = render((
      <div>
        <Input form="test" name="email" invalidMessage="error" isRequired validator="email" value={email} />
        <Button onClick={onSubmitValid} onValidationFail={onSubmitInvalid} form="test">test</Button>
      </div>
    ));

    fireEvent.click(getByText('test'));
    expect(onSubmitValid).toBeCalledTimes(0);
    expect(onSubmitInvalid).toBeCalledTimes(1);
  });
});
