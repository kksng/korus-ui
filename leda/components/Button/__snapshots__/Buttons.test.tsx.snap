// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshots collection is Button render right with Form and onClick props? 1`] = `
<div>
  <div>
    <div
      class="input-wrapper"
    >
      <div
        class="input-element-wrapper required"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="input-element"
          form="test"
          name="in"
          value=""
        />
      </div>
    </div>
    <button
      class="button-wrapper disabled"
      type="button"
    >
      test
    </button>
  </div>
</div>
`;

exports[`Button snapshots collection is Button render right with basic props? 1`] = `
<div>
  <button
    class="button-wrapper"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Button snapshots collection is Button render right with isDisabled attributes? 1`] = `
<div>
  <button
    class="button-wrapper disabled"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Button snapshots collection is Button render right with isLoading? 1`] = `
<div>
  <button
    class="button-wrapper loading"
    type="button"
  >
    test
  </button>
</div>
`;
